### PR TITLE
coq_HEAD: Update to the latest commit

### PR DESCRIPTION
**Pull request by rht**
###### Motivation for this change
Testing an automatically generated nix file update, through https://github.com/rht/nixpkg-git-fetch-head/blob/master/nixpkg-git-fetch-head.sh.
With tiny modification (e.g. replace the ls-remote HEAD tag with release tags), this could be used for other packages that mainly use `fetchgit`.

###### Things done

- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [x] Linux
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by vbgl** [comment]
The value of the `version` attribute looks weird. In particular it makes the derivation name `coq-V8.6beta1` and version `203-g…`; none of which makes much sense to me.

**Comment by rht** [comment]
What should be a consistent way to name the `version` such that this can be specified in the code? It is generated by `git-describe` as has been used in `fetchGit`. `203` is the number of tags present in the coq repo, and can be clipped away. Would `coq-v8.6beta1` and `g37817bb5a` be fine?

**Comment by vbgl** [comment]
`coq-V8.6beta1` is definitely a wrong name: `coq` is enough, or `coq-head` (or `coq-master` or whatever) if you want to distinguish is from the released ones. For the version you need a number that increases with time (before the hash); the commit date might be more readable than the number of tags.

Finally, what about something like `coq-head-${date}-${git-describe}`?

**Comment by rht** [comment]
I just checked from https://git-scm.com/docs/git-describe that the number represents the number of additional commits since the most recent tag. The prefix 'g' refers to git. Have yet to check how other vcs describe the version of their most recent commits.

**Comment by rht** [comment]
@vbgl how should I put a constraint on the ocaml version s.t. it is at least 4.02.1?

**Comment by rht** [comment]
> For the version you need a number that increases with time (before the hash); the commit date might be more readable than the number of tags.

This is simpler if it is up to the standard convention of each vcs.

**Comment by vbgl** [comment]
What do you call a constraint? If you expect nix to solve such constraints then I don’t know.

You can check the property you mention with `stdenv.lib.versionAtLeast ocaml.version "4.02.1"`. If this evaluates to false, you can return `null`, throw an error, return a derivation with a `meta.broken` attribute set to true, etc.

**Comment by rht** [comment]
I see, tho is it possible to specify something along the line of "4.02.1" then have nix automatically overrides the default if it is older than that, rather than having a broken package? Currently, I have to put this by hand at `./pkgs/top-level/all-packages.nix`.

**Comment by rht** [comment]
I think the only objection remaining for this PR is the versioning scheme, which as I stated, would be better to reuse the existing tag by the coq package maintainer (the number of extra commits rather than the timestamp).

**Comment by vbgl** [comment]
Let me rephrase: the version number should start by a digit; this is why I suggest to put a timestamp in front of the output of `git-describe`. Otherwise, this changes the name of the package to something rather confusing.

**Comment by rht** [comment]
I see, it looks like there is one case e.g. `noto-fonts` with version `git-2016-03-29`. I have added the date, now it is `2017-01-09-V8.6beta1-203-g37817bb5a`

**Comment by rht** [comment]
alt:`coq-git-V8.6beta1-203-g37817bb5a`

File: pkgs/applications/science/logic/coq/HEAD.nix [hunk 1]
```diff
@@ -1,26 +1,29 @@
...
   version = "2017-01-09-V8.6beta1-203-g37817bb5a";
   coq-version = "8.6";
-  buildIde = lablgtk != null;
-  ideFlags = if buildIde then "-lablgtkdir ${lablgtk}/lib/ocaml/*/site-lib/lablgtk2 -coqide opt" else "";
+  ideFlags = if buildIde then "-lablgtkdir ${ocamlPackages_4_02.lablgtk}/lib/ocaml/*/site-lib/lablgtk2 -coqide opt" else "";
   csdpPatch = if csdp != null then ''
     substituteInPlace plugins/micromega/sos.ml --replace "; csdp" "; ${csdp}/bin/csdp"
     substituteInPlace plugins/micromega/coq_micromega.ml --replace "System.is_in_system_path \"csdp\"" "true"
   '' else "";
+  ocaml = ocamlPackages_4_02.ocaml;
```
Comment (rht) [inline pkgs/applications/science/logic/coq/HEAD.nix]: @vbgl I overrode the definition here so that there is no need to put an extra prefix `ocamlPackages_4_02` in `buildInputs`. Plus, the version-specific statement is localized to only within the let clause. WDYT?

File: pkgs/applications/science/logic/coq/HEAD.nix [hunk 2]
```diff
@@ -1,31 +1,34 @@
...
+  lablgtk = ocamlPackages_4_02.lablgtk;
+  camlp5 = ocamlPackages_4_02.camlp5_transitional;
 in
 
 stdenv.mkDerivation {
   name = "coq-${version}";
 
   inherit coq-version;
-  inherit ocaml camlp5;
+  inherit ocaml findlib lablgtk camlp5;
```
Comment (vbgl) [inline pkgs/applications/science/logic/coq/HEAD.nix]: What’s the purpose of this change?

Comment (rht) [inline pkgs/applications/science/logic/coq/HEAD.nix]: There are several ways to specify the scope (in all-packages.nix, in the let clause, or casted in inherit and buildInputs), I picked the `let` option because it is most localized within each packages nix file. If it were to be casted in inherit and buildInputs it'd be too verbose.

Comment (vbgl) [inline pkgs/applications/science/logic/coq/HEAD.nix]: This change adds two attributes to the argument of `mkDerivation` these two attributes are never used. So I wonder why you define them.

The `ocaml` and `camlp5` attributes are not used it this derivation but in some Coq libraries that define OCaml plugins: they have to be compiled with the same version of the OCaml compiler that is used to compile Coq itself. That’s why these attributes “remember” with which OCaml/camlp5 Coq is compiled.

BTW, what do you call “casted”?

Comment (rht) [inline pkgs/applications/science/logic/coq/HEAD.nix]: > This change adds two attributes to the argument of mkDerivation these two attributes are never used. So I wonder why you define them.

I assumed this is necessary because of https://github.com/NixOS/nixpkgs/pull/21822/files#diff-036410e9211b4336186fc613f7200b12L16966. But it looks like defining it in `let` is sufficient for coq-modules to get the right version of camlp5.

> BTW, what do you call “casted”?
I refer `(ocamlPackages) ocaml ...` as casting.

Comment (rht) [inline pkgs/applications/science/logic/coq/HEAD.nix]: Confirmed, `coq.camlp5` has been used several times but neither for `coq.findlib` and `coq.lablgtk`.

Comment (vbgl) [inline pkgs/applications/science/logic/coq/HEAD.nix]: > I refer `(ocamlPackages) ocaml ...` as casting.

This piece of syntax cannot exist alone: it must come after the `inherit` keyword. It has nothing to do with type casting; it defines some attributes. See [the nix manual](http://nixos.org/nix/manual/#idm140737318075520) for details.

Comment (rht) [inline pkgs/applications/science/logic/coq/HEAD.nix]: I know, I was just using 'casting' to describe the parentheses syntax. From the manual, this wasn't stated explicitly, only as an example `(xlibs)`.

**Comment by vbgl** [comment]
> I know, I was just using 'casting' to describe the parentheses syntax.

If you really want to compare with C or Java syntax, it is much more
similar to an `if` construct (as in `if (b) ++x`; from this snippet, it
would make no sense to refer only to the `(b) ++x` fragment).

Back to the topic of this PR, the Nixpkgs contributor guide has [clear
recommendations on how to name
packages](https://nixos.org/nixpkgs/manual/#sec-package-naming),
specially when packaging “not a release but a commit from a repository”.
So please prepend `unstable-` to the version attribute (or insert it in
the name attribute).

**Comment by rht** [comment]
Well, fine, in C-like langs, typecasting / type conversion can't be stated as a stand-alone expression, while `++` could.

Since version MUST start with a number, so I suppose it should inserted into the name attribute.
The point of this PR is to update the coq_HEAD version (making the package name properly formed, is a side effect).

**Comment by rht** [comment]
(This is outside the scope of this PR, but any conclusion/standardization made here should apply to these packages as well:
```
./pkgs/applications/editors/emacs-modes/proofgeneral/HEAD.nix
./pkgs/development/coq-modules/fiat/HEAD.nix
./pkgs/development/libraries/xcb-util-cursor/HEAD.nix
./pkgs/development/compilers/ghc/head.nix
./pkgs/development/compilers/ghcjs/head.nix
```)

**Comment by vbgl** [comment]
> Confirmed, `coq.camlp5` has been used several times but neither for `coq.findlib` and `coq.lablgtk`.

Now there is a case for `coq.findlib` in ssreflect and mathcomp.

**Comment by rht** [comment]
`coq.findlib` added.

File: pkgs/applications/networking/remote/freerdp/unstable.nix [hunk 3]
```diff
@@ -6,6 +6,7 @@
 }:
 
 stdenv.mkDerivation rec {
+  version = ";";
```
Comment (globin) [inline pkgs/applications/networking/remote/freerdp/unstable.nix]: Why was this line added?

Comment (rht) [inline pkgs/applications/networking/remote/freerdp/unstable.nix]: (accidentally did `git commit -am` on an unstaged local change ... reverting this line)
